### ISSUE TITLE
Mod cgrates not terminating session upon error or dialog failure

### ISSUE THREAD

**Issue by acheraime** [OpenSIPS/opensips#2380]
Good morning all,

I'm experiencing an issue with cgrates integration where opensips will not instruct Cgrates to terminate a session when an error happens or a dialog fail due to missing ACK or something else.
On Cgrates side, the session will remain active until manually terminated or expired.

**OpenSIPS version**
```
version: opensips 3.0.5 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 7c92fff
main.c compiled on 23:34:14 Jan 12 2021 with gcc 4.8.5
```

**Describe the bug**
Mod cgrates will not communicate back to Cgrates Session manager when there's an error or a dialog misbehavior by example when dialog fail with code 3 (missing ACK).
In such a situation, the cgrates session will remain hang and accounting will keep going until that specific account balance is depleted to zero or the ttl is reached whichever comes first.

**Debug**
```
Jan 26 00:19:04 [6481] DBG:tm:run_trans_callbacks: trans=0x7efd787cb428, callback type 128, id 4 entered
Jan 26 00:19:04 [6481] DBG:cgrates:cgr_tmcb_func: Called callback for transaction 0x7efd787cb428 type 128 reply_code=200 branch=0
Jan 26 00:19:04 [6481] DBG:cgrates:cgr_handle_cmd: sending json string: { "method": "SessionSv1.InitiateSession", "id": 2088663509, "params": [ { "Category": "call", "Tenant": "xxxxxx", "ProcessStatQueues": true, "ProcessThresholds": true, "ReleaseResources": false, "AllocateResources": false, "AuthorizeResources": false, "GetMaxUsage": true, "Event": { "Source": "edge-proxy-1", "OriginHost": "10.20.100.2", "RequestType": "*prepaid", "OriginID": "d7c7a772-da38-1239-29ad-42010a146403", "DialogID": 808309049, "DialogEntry": 1232, "Account": "71e4335cxxxxxxxxxxxxxxxx", "SetupTime": "1611638342", "AnswerTime": "1611638344", "Destination": "5xxxxxxxxxxxxxxx" }, "InitSession": true } ] }
Jan 26 00:19:04 [6481] DBG:cgrates:cgr_get_default_conn: using default connection - running in sync mode!
Jan 26 00:19:04 [6481] DBG:cgrates:cgrc_send: Successfully sent 626 bytes
Jan 26 00:19:04 [6481] DBG:cgrates:cgrc_async_read: Event on fd 110 from 10.22.100.2:2014
Jan 26 00:19:04 [6481] DBG:cgrates:cgrc_async_read: **Received (possible partial) json**: {{"id":2088663509,"result":{"Attributes":null,"ResourceAllocation":null,"MaxUsage":2160000000000,"ThresholdIDs":null,"StatQueueIDs":null},"error":null}
}
Jan 26 00:19:04 [6481] DBG:cgrates:cgrates_process: Processing JSON-RPC: { "id": 2088663509, "result": { "Attributes": null, "ResourceAllocation": null, "MaxUsage": 2160000000000, "ThresholdIDs": null, "StatQueueIDs": null }, "error": null }
Jan 26 00:19:04 [6481] DBG:cgrates:cgrates_process: **treating JSON-RPC as a reply**
Jan 26 00:19:04 [6481] DBG:dialog:lookup_dlg: dialog id=808309049 found on entry 1232
Jan 26 00:19:04 [6481] DBG:dialog:init_dlg_term_reason: Setting DLG term reason to [CGRateS Accounting Denied]
Jan 26 00:19:04 [6481] DBG:dialog:send_leg_bye: sending BYE on dialog 0x7efd787cecf8 to caller (0)

```

**To Reproduce**
To reproduce this issue, one needs to have Opensips integrated with Cgrates. 
1. Start a session in *prepaid mode
2. Call cgrates_auth
3. Uppon successful authentication and authorization start accounting with cgrates_acc
4. Simulate a missing ACK from the dialog
5. Observe that the call will terminate with a dialog end reason of "Cgrates accounting denied"
6. On cgrates side check the account balance
7. On cgrates cgr-console execute "active_sessions" you should see the session being active

**Expected behavior**
The expectation is that when dialog error or any related error happens, Opensips should send 
a session terminate (SessionSv1.TerminateSession) request to Cgrates.

**Comment by razvancrainea** [comment]
Is this still happening? I've just tested this and the `TerminateSession` command is properly sent downstream.
Moreover, the `CGRateS Accounting Denied` is set due to a different cause, mainly due to a CGRateS deny, not due to a missing ACK.

**Comment by vtzan** [comment]
Dear @acheraime @razvancrainea,

I have a testing setup with opensips + cgrates. @acheraime could you please provide detail information about how you are simulating the missing ACK from dialog so that we can be sure that we test the same scenario? @razvancrainea any idea what may @acheraime means?

Best Regards
Vasilios Tzanoudakis

**Comment by acheraime** [comment]
@vtzan as @razvancrainea mentioned this has nothing to do with missing ACK. It was an issue with the module not correctly parsing the MaxUsage token value received in the response payload from CGRateS.

**Comment by acheraime** [comment]
> Is this still happening? I've just tested this and the `TerminateSession` command is properly sent downstream.
> Moreover, the `CGRateS Accounting Denied` is set due to a different cause, mainly due to a CGRateS deny, not due to a missing ACK.

Your patch has fixed the issue but now it seems like the dialog is not terminated. The call will go on until one of the legs send the BYE which cause the account balance to go negative. I ensure that a dialog is created with the "B" flag set in the initial INVITE.

Thanks for your unconditional support.

Adolphe Cher-Aime

**Comment by razvancrainea** [comment]
Are you somehow/somehwere in your script setting the `$DLG_timeout variable`? Can you track down what is the MaxUsage value, and see if the dialog's timeout is properly adjusted (run `opensips-cli -x mi dlg_list`) and check the timeout values.

**Comment by acheraime** [comment]
> opensips-cli -x mi dlg_list

Hi @razvancrainea ,

The $DLG_timeout is not set anywhere in my script. Yes MaxUsage value is set appropriately as returned by CGRateS here
"**setting dialog timeout to 660**"
```bash
          Apr  2 09:41:52 edge-proxy-1 /usr/local/sbin/opensips[25198]: DBG:cgrates:cgrates_process: Processing JSON-RPC: { 
         "id": 1158057354, "result": { "Attributes": null, "ResourceAllocation": null, "MaxUsage": 660000000000, 
         "ThresholdIDs": null, "StatQueueIDs": null }, "error": null }
          Apr  2 09:41:52 edge-proxy-1 /usr/local/sbin/opensips[25198]: DBG:cgrates:cgrates_process: treating JSON-RPC as a 
          reply
          Apr  2 09:41:52 edge-proxy-1 /usr/local/sbin/opensips[25198]: DBG:cgrates:cgr_proc_start_acc_reply: setting dialog 
         timeout to 660
```
 
However, the dialog list shows the timeout is set to the default value of 30 minutes.
**"datestart": "2021-04-02 09:41:52"**
**"dateout": "2021-04-02 10:11:52"**

```bash
             "ID": "64d.029dbfd4",
            "db_id": "14595607222560",
            "state": 4,
            "user_flags": 0,
            "timestart": 1617370912,
            "datestart": "2021-04-02 09:41:52",
            "timeout": 1617372712,
            "dateout": "2021-04-02 10:11:52",
            "callid": "04ad4175-0e5c-123a-29ad-42010a146403",
            "from_uri": "sip:+1xxxxxxxxx08@10.20.100.3",
            "to_uri": "sip:011xxxxxxxxxx00@10.20.100.2",
            "caller_tag": "BS8Z6HyKX4U1g",
            "caller_contact": "sip:mod_sofia@10.20.100.3:5080",
            "callee_cseq": "0",
            "caller_route_set": "",
            "caller_bind_addr": "udp:10.20.100.2:5060",
```

Best Regards,

Adolphe Cher-Aime

**Comment by acheraime** [comment]
@razvancrainea,

Any update on this?

Best,
Acheraime

**Comment by razvancrainea** [comment]
I've just pushed a possible fix for this issue in the master branch. Any chance you could give it a try?

Best regards,
Răzvan

**Comment by acheraime** [comment]
> I've just pushed a possible fix for this issue in the master branch. Any chance you could give it a try?
> 
> Best regards,
> Răzvan
Hi Razvan,

Thank you for your hard work. I am not ready for 3.2 yet. Any possibility to get this change into 3.1?


Best,

Acheraime

**Comment by vtzan** [comment]
Yes please backport it to 3.1

Thanks

---
Vasilios Tzanoudakis

Στις Παρ, 16 Απρ 2021, 16:00 ο χρήστης Adolphe Cher-Aime <
***@***.***> έγραψε:

**Comment by razvancrainea** [comment]
Sure, backporting it to 3.1.

**Comment by acheraime** [comment]
Thanks @razvancrainea. I have tested unfortunately I see the same behavior. The dialog timeout is set to the default value configured in the module.
@vtzan did you have a chance to test it?

Regards,

Acheraime

**Comment by vtzan** [comment]
@acheraime not yet but I will test it soon and share the results from my side too ;-)

**Comment by vtzan** [comment]
Dear @razvancrainea @acheraime,

While doing some quick tests I am sharing the results : 

My default timeout:
modparam("dialog", "default_timeout", 100)

1)Make a call
2)Cgrates sends "MaxUsage": 10800 to --> Opensips
3)OpenSIPS returns $cgr_ret(MaxUsage)=1  (why?)
4)And OpenSIPS dlg_list has the following :
------------------------------------------------------
"timestart": 1618594461,
"datestart": "2021-04-16 17:34:21",
"timeout": 1618594561,
"dateout": "2021-04-16 17:36:01",
------------------------------------------------------

So yes I think the problem is still there but could you  also explain $cgr_ret(MaxUsage)=1?

thank you in advance

Vasilios Tzanoudakis

**Comment by acheraime** [comment]
> Dear @razvancrainea @acheraime,
> 
> While doing some quick tests I am sharing the results :
> 
> My default timeout:
> modparam("dialog", "default_timeout", 100)
> 
> ## 1)Make a call
> 2)Cgrates sends "MaxUsage": 10800 to --> Opensips
> 3)OpenSIPS returns $cgr_ret(MaxUsage)=1 (why?)
> 4)And OpenSIPS dlg_list has the following :
> ## "timestart": 1618594461,
> "datestart": "2021-04-16 17:34:21",
> "timeout": 1618594561,
> "dateout": "2021-04-16 17:36:01",
> So yes I think the problem is still there but could you also explain $cgr_ret(MaxUsage)=1?
> 
> thank you in advance
> 
> Vasilios Tzanoudakis

@vtzan about the MaxUsage, it's been like that for quite a longtime. What I had to do is to parse the json response and use the MaxUsage token to retrieve the value.
```bash
$json(cgrRes) := $cgr_ret;
$var(maxUsage) = $json(cgrRes/MaxUsage);
```

**Comment by razvancrainea** [comment]
Any chance I could get some logs for this?
@acheraime are you sure you were using the latest code, that included the fix? Can you send the output of `opensips -V`?

**Comment by acheraime** [comment]
> Any chance I could get some logs for this?
> @acheraime are you sure you were using the latest code, that included the fix? Can you send the output of `opensips -V`?

Hi @razvancrainea sorry for the delay. I was out, please  see version below and the log in this pastebin.


https://pastebin.com/LnysRmMv

```bash
version: opensips 3.1.1 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: fda62f9
main.c compiled on 07:58:08 Apr 19 2021 with gcc 4.8.5
```

Best,

--
Acheraime

**Comment by razvancrainea** [comment]
@acheraime, in the logs you provided I cannot see the request sent to CGrateS, only the reply. However, I presume it is a `SessionSv1.InitiateSession` command - can you please confirm?
However, after the reply is received, I see that OpenSIPS processes a request, rather than a 200 OK reply - so what I can presume is that you are sending the command manually, rather than using the `cgrates_acc` function.
Note that CGRateS returns a `3240ns` max usage, which divided by `1000000000` results in `0s`, thus the dialog should have been immediately terminated. But since you're probably not going with the `cgrates_acc` function, you should implement the logic yourself.
Bottom line, I am not seeing any issues here, as I don't have enough information, and the one that I have does not indicate any problems.
If you can provide a full debug for an entire call flow, I can re-assess this.

Best regards,
Răzvan
